src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp

```diff
@@ -490,6 +495,105 @@ void RenderingManager::slot_camera_change_end()
+        return;
+    }
+
+    DroppedMaterialMenu* menu = new DroppedMaterialMenu(drop_pos, material_name);
+
+    connect(menu,
+        SIGNAL(signal_apply_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)),
+        SLOT(slot_change_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)));
+
+    menu->exec(QCursor::pos());
+}
+
+namespace
+{
+    struct MaterialSlot
+    {
+        std::string name;
+        ObjectInstance::Side side;
+    };
+
+    class ApplyDroppedMaterial
+        : public RenderingManager::IScheduledAction
+    {
+    public:
+        explicit ApplyDroppedMaterial(
+            ObjectInstance& instance,
+            const std::vector<MaterialSlot>& material_slots,
+            const std::string& material_name)
+            : m_instance(instance),
+            m_material_slots(material_slots),
+            m_material_name(material_name)
+        {
+        }
+
+        void operator()(Project& project) override
+        {
+            for (const_each<std::vector<MaterialSlot>> slot = m_material_slots; slot; ++slot)
+            {
+                m_instance.assign_material(slot->name.c_str(), slot->side, m_material_name.c_str());
+            }
+        }
+
+    private:
+        ObjectInstance& m_instance;
+        const std::vector<MaterialSlot> m_material_slots;
+        const std::string m_material_name;
+    };
+}
+
+void RenderingManager::slot_change_material(
+    const foundation::Vector2d& drop_pos,
+    const std::string& material_name,
+    renderer::ObjectInstance::Side side)
+{
+    const ScenePicker scene_picker(*m_project);
+    const ScenePicker::PickingResult result = scene_picker.pick(drop_pos);
+    renderer::ObjectInstance* instance = const_cast<renderer::ObjectInstance*>(result.m_object_instance);
+
+    std::vector<MaterialSlot> material_slots;
+    if (side == ObjectInstance::Side::FrontSide || side == ObjectInstance::Side::BothSides)
+    {
+        for (auto item = instance->get_front_material_mappings().begin();
+            item != instance->get_front_material_mappings().end();
+            ++item)
+        {
+            material_slots.push_back({ item.key(), ObjectInstance::Side::FrontSide });
+        }
+    }
+    if (side == ObjectInstance::Side::BackSide || side == ObjectInstance::Side::BothSides)
+    {
+        for (auto item = instance->get_back_material_mappings().begin();
+            item != instance->get_back_material_mappings().end();
+            ++item)
+        {
+            material_slots.push_back({ item.key(), ObjectInstance::Side::BackSide });
+        }
+    }
+
+    schedule_or_execute(unique_ptr<RenderingManager::IScheduledAction>(
+        new ApplyDroppedMaterial(
+            *instance,
+            material_slots,
+            material_name)));
+
+    if (is_rendering())
```

Comment:
Redundant I think, this is already done by `schedule_or_execute()`.

Reply:
I took for example materialassigmenteditorwindow.cpp L367
`    m_editor_context.m_rendering_manager.schedule_or_execute(
        unique_ptr<RenderingManager::IScheduledAction>(
            new AssignMaterialsAction(
                m_object_instance,
                m_object_instance_item,
                slot_values)));

    if (old_front_mappings != m_object_instance.get_front_material_mappings() ||
        old_back_mappings != m_object_instance.get_back_material_mappings())
        m_editor_context.m_project_builder.slot_notify_project_modification();

    if (m_editor_context.m_rendering_manager.is_rendering())
        m_editor_context.m_rendering_manager.reinitialize_rendering();`
Do you want that I change this one too?

Reply:
Yeah, that was a mistake, it seems that we need to fix that too. Please make sure it continues to work afterward (i.e. open the material assignment window and assign materials _during_ rendering).

Reply:
It is really weird, it seems assign material apply the material 2 times, when you select the material and when you apply it
can you confirm it does the same on your version?

Reply:
it seems it is the wanted behavior

Reply:
Yes, it's the desired behavior, the first assignment is caused by this code:

```
void MaterialAssignmentEditorWindow::slot_entity_browser_accept(
    QLineEdit*          line_edit,
    QString             page_name,
    QString             entity_name)
{
    line_edit->setText(entity_name);

	// HERE
    assign_materials(get_slot_values());

    qobject_cast<QWidget*>(sender()->parent())->close();
}
```